File: chtl-compiler/src/chtl/lexer/lexer.rs
use crate::chtl::config_manager::ConfigManager;
use crate::chtl::lexer::token::Token;
use std::iter::Peekable;
use std::str::Chars;

pub struct Lexer<'a> {
    input: Peekable<Chars<'a>>,
    ch: char,
    config: &'a ConfigManager,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str, config: &'a ConfigManager) -> Self {
        let mut l = Lexer {
            input: input.chars().peekable(),
            ch: ' ',
            config,
        };
        l.read_char();
        l
    }

    fn read_char(&mut self) {
        self.ch = self.input.next().unwrap_or('\0');
    }

    pub fn next_token(&mut self) -> Token {
        self.skip_whitespace();

        let tok = match self.ch {
            '=' | ':' => Token::Colon,
            ';' => Token::Semicolon,
            '?' => Token::Question,
            '(' => Token::LParen,
            ')' => Token::RParen,
            '{' => Token::LBrace,
            '}' => Token::RBrace,
            '[' => Token::LBracket,
            ']' => Token::RBracket,
            ',' => Token::Comma,
            '.' => Token::Dot,
            '+' => Token::Plus,
            '-' => Token::Minus,
            '/' => {
                if let Some(&c) = self.input.peek() {
                    if c == '/' {
                        self.read_char(); // consume first /
                        self.read_char(); // consume second /
                        return Token::SingleLineComment(self.read_single_line_comment());
                    } else if c == '*' {
                        self.read_char(); // consume /
                        self.read_char(); // consume *
                        return Token::MultiLineComment(self.read_multi_line_comment());
                    }
                }
                Token::Slash
            }
            '*' => {
                if let Some(&'*') = self.input.peek() {
                    self.read_char();
                    Token::Power
                } else {
                    Token::Asterisk
                }
            }
            '%' => Token::Percent,
            '>' => Token::Gt,
            '<' => Token::Lt,
            '!' => Token::Bang,
            '&' => {
                if let Some(&'&') = self.input.peek() {
                    self.read_char();
                    Token::And
                } else {
                    Token::Ampersand
                }
            }
            '|' => {
                if let Some(&'|') = self.input.peek() {
                    self.read_char();
                    Token::Or
                } else {
                    Token::Illegal
                }
            }
            '@' => Token::At,
            '#' => {
                if let Some(&' ') = self.input.peek() {
                    self.read_char(); // consume #
                    self.read_char(); // consume space
                    return Token::GeneratorComment(self.read_single_line_comment());
                } else {
                    Token::Hash
                }
            }
            '"' | '\'' => {
                return self.read_string();
            }
            '$' => {
                return self.read_responsive_value();
            }
            '\0' => Token::Eof,
            _ => {
                if is_letter(self.ch) {
                    let literal = self.read_identifier();
                    return self.lookup_ident(&literal);
                } else if is_digit(self.ch) {
                    return self.read_number();
                } else {
                    Token::Illegal
                }
            }
        };

        self.read_char();
        tok
    }

    fn skip_whitespace(&mut self) {
        while self.ch.is_whitespace() {
            self.read_char();
        }
    }

    fn read_identifier(&mut self) -> String {
        let mut identifier = String::new();
        while is_letter(self.ch) || is_digit(self.ch) {
            identifier.push(self.ch);
            self.read_char();
        }
        identifier
    }

    fn read_number(&mut self) -> Token {
        let mut number = String::new();
        while is_digit(self.ch) || self.ch == '.' {
            number.push(self.ch);
            self.read_char();
        }

        let unit = self.read_unit();

        Token::Number(number, unit)
    }

    fn read_unit(&mut self) -> Option<String> {
        // CSS units can contain letters or a percentage sign.
        // We'll keep it simple and read any sequence of letters.
        if is_letter(self.ch) || self.ch == '%' {
            let mut unit = String::new();
            while is_letter(self.ch) || self.ch == '%' {
                unit.push(self.ch);
                self.read_char();
            }
            return Some(unit);
        }
        None
    }

    fn read_string(&mut self) -> Token {
        let quote_type = self.ch;
        let mut value = String::new();
        self.read_char(); // Consume opening quote
        while self.ch != quote_type && self.ch != '\0' {
            value.push(self.ch);
            self.read_char();
        }
        self.read_char(); // Consume closing quote
        Token::String(value)
    }

    fn read_responsive_value(&mut self) -> Token {
        self.read_char(); // Consume opening '$'
        let mut identifier = String::new();
        while self.ch != '$' && self.ch != '\0' {
            identifier.push(self.ch);
            self.read_char();
        }
        self.read_char(); // Consume closing '$'
        Token::ResponsiveValue(identifier)
    }

    fn read_single_line_comment(&mut self) -> String {
        let mut comment = String::new();
        while self.ch != '\n' && self.ch != '\0' {
            comment.push(self.ch);
            self.read_char();
        }
        comment
    }

    fn read_multi_line_comment(&mut self) -> String {
        let mut comment = String::new();
        while !(self.ch == '*' && self.input.peek() == Some(&'/')) {
            if self.ch == '\0' {
                return comment;
            }
            comment.push(self.ch);
            self.read_char();
        }
        self.read_char(); // consume '*'
        self.read_char(); // consume '/'
        comment
    }

    fn lookup_ident(&self, ident: &str) -> Token {
        self.config
            .keyword_tokens
            .get(ident)
            .cloned()
            .unwrap_or_else(|| Token::Identifier(ident.to_string()))
    }

    pub fn read_raw_body(&mut self) -> String {
        let mut content = String::new();
        let mut brace_level = 1;

        while brace_level > 0 && self.ch != '\0' {
            if self.ch == '{' {
                brace_level += 1;
            } else if self.ch == '}' {
                brace_level -= 1;
                if brace_level == 0 {
                    break;
                }
            }
            content.push(self.ch);
            self.read_char();
        }
        self.read_char();
        content
    }
}

fn is_letter(ch: char) -> bool {
    ch.is_alphabetic() || ch == '_' || ch == '-'
}

fn is_digit(ch: char) -> bool {
    ch.is_ascii_digit()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::chtl::lexer::token::Token;

    use crate::chtl::node::ast::Statement;
    use crate::chtl::parser::parser::Parser;

    #[test]
    fn test_custom_keyword_from_config() {
        let config_input = r#"
        [configuration] {
            [name] {
                KEYWORD_IF: "my_if";
            }
        }
        "#;
        let mut config = ConfigManager::new();
        let config_lexer = Lexer::new(config_input, &config);
        let mut config_parser = Parser::new(config_lexer);
        let config_program = config_parser.parse_program();
        if let Some(Statement::Configuration(config_stmt)) = config_program.statements.get(0) {
            config.apply_config(config_stmt);
        } else {
            panic!("Failed to parse configuration statement");
        }

        let chtl_input = "my_if condition;";
        let mut lexer = Lexer::new(chtl_input, &config);

        let first_token = lexer.next_token();
        assert_eq!(first_token, Token::If);

        let second_token = lexer.next_token();
        assert_eq!(second_token, Token::Identifier("condition".to_string()));
    }

    #[test]
    fn test_next_token() {
        let input = r#"
        div {
            id: "box";
            class: welcome;
            text { "Hello World" }
        }

        // This is a comment
        /* This is a multi-line comment */
        # This is a generator comment

        style {
            .box {
                width: 100px;
                height: 200.5px;
            }
        }
        "#;

        let tests = vec![
            Token::Identifier("div".to_string()),
            Token::LBrace,
            Token::Identifier("id".to_string()),
            Token::Colon,
            Token::String("box".to_string()),
            Token::Semicolon,
            Token::Identifier("class".to_string()),
            Token::Colon,
            Token::Identifier("welcome".to_string()),
            Token::Semicolon,
            Token::Text,
            Token::LBrace,
            Token::String("Hello World".to_string()),
            Token::RBrace,
            Token::RBrace,
            Token::SingleLineComment(" This is a comment".to_string()),
            Token::MultiLineComment(" This is a multi-line comment ".to_string()),
            Token::GeneratorComment("This is a generator comment".to_string()),
            Token::Style,
            Token::LBrace,
            Token::Dot,
            Token::Identifier("box".to_string()),
            Token::LBrace,
            Token::Identifier("width".to_string()),
            Token::Colon,
            Token::Number("100".to_string(), Some("px".to_string())),
            Token::Semicolon,
            Token::Identifier("height".to_string()),
            Token::Colon,
            Token::Number("200.5".to_string(), Some("px".to_string())),
            Token::Semicolon,
            Token::RBrace,
            Token::RBrace,
            Token::Eof,
        ];

        let config = ConfigManager::new();
        let mut lexer = Lexer::new(input, &config);

        for expected_token in tests {
            let tok = lexer.next_token();
            assert_eq!(tok, expected_token);
        }
    }

    #[test]
    fn test_ce_equivalence() {
        let input = "id: box; class = container;";
        let tests = vec![
            Token::Identifier("id".to_string()),
            Token::Colon,
            Token::Identifier("box".to_string()),
            Token::Semicolon,
            Token::Identifier("class".to_string()),
            Token::Colon,
            Token::Identifier("container".to_string()),
            Token::Semicolon,
            Token::Eof,
        ];

        let config = ConfigManager::new();
        let mut lexer = Lexer::new(input, &config);
        for expected_token in tests {
            let tok = lexer.next_token();
            assert_eq!(tok, expected_token);
        }
    }

    #[test]
    fn test_text_keyword() {
        let input = "text { \"some text\" }";
        let tests = vec![
            Token::Text,
            Token::LBrace,
            Token::String("some text".to_string()),
            Token::RBrace,
            Token::Eof,
        ];

        let config = ConfigManager::new();
        let mut lexer = Lexer::new(input, &config);
        for expected_token in tests {
            let tok = lexer.next_token();
            assert_eq!(tok, expected_token);
        }
    }

    #[test]
    fn test_percentage_unit() {
        let input = "width: 50%;";
        let tests = vec![
            Token::Identifier("width".to_string()),
            Token::Colon,
            Token::Number("50".to_string(), Some("%".to_string())),
            Token::Semicolon,
            Token::Eof,
        ];

        let config = ConfigManager::new();
        let mut lexer = Lexer::new(input, &config);
        for expected_token in tests {
            let tok = lexer.next_token();
            assert_eq!(tok, expected_token);
        }
    }

    #[test]
    fn test_import_keywords() {
        let input = r#"
        [import] @html from "a.html" as A;
        [import] @javascript from "b.js" as B;
        [import] @chtl from "c.chtl" as C;
        [import] @cjmod from "d.cjmod" as D;
        [import] @config from "e.chtl" as E;
        "#;

        let tests = vec![
            // First line
            Token::LBracket,
            Token::Import,
            Token::RBracket,
            Token::At,
            Token::Html,
            Token::From,
            Token::String("a.html".to_string()),
            Token::As,
            Token::Identifier("A".to_string()),
            Token::Semicolon,
            // Second line
            Token::LBracket,
            Token::Import,
            Token::RBracket,
            Token::At,
            Token::JavaScript,
            Token::From,
            Token::String("b.js".to_string()),
            Token::As,
            Token::Identifier("B".to_string()),
            Token::Semicolon,
            // Third line
            Token::LBracket,
            Token::Import,
            Token::RBracket,
            Token::At,
            Token::Chtl,
            Token::From,
            Token::String("c.chtl".to_string()),
            Token::As,
            Token::Identifier("C".to_string()),
            Token::Semicolon,
            // Fourth line
            Token::LBracket,
            Token::Import,
            Token::RBracket,
            Token::At,
            Token::CjMod,
            Token::From,
            Token::String("d.cjmod".to_string()),
            Token::As,
            Token::Identifier("D".to_string()),
            Token::Semicolon,
            // Fifth line
            Token::LBracket,
            Token::Import,
            Token::RBracket,
            Token::At,
            Token::Config,
            Token::From,
            Token::String("e.chtl".to_string()),
            Token::As,
            Token::Identifier("E".to_string()),
            Token::Semicolon,
            Token::Eof,
        ];

        let config = ConfigManager::new();
        let mut lexer = Lexer::new(input, &config);

        for expected_token in tests {
            let tok = lexer.next_token();
            assert_eq!(tok, expected_token);
        }
    }

    #[test]
    fn test_logical_operators() {
        let input = "&& ||";
        let tests = vec![Token::And, Token::Or];
        let config = ConfigManager::new();
        let mut lexer = Lexer::new(input, &config);
        for expected_token in tests {
            let tok = lexer.next_token();
            assert_eq!(tok, expected_token);
        }
    }

    #[test]
    fn test_case_sensitivity() {
        let input = "If IF if";
        let config = ConfigManager::new();
        let mut lexer = Lexer::new(input, &config);

        let tests = vec![
            Token::Identifier("If".to_string()),
            Token::Identifier("IF".to_string()),
            Token::If,
        ];

        for expected_token in tests {
            let tok = lexer.next_token();
            assert_eq!(tok, expected_token);
        }
    }
}

File: CHTL/src/lexer/Lexer.cpp
#include "lexer/Lexer.h"
#include <cctype>

namespace CHTL
{
    Lexer::Lexer(const std::string& input, const std::unordered_map<std::string, Token>& keywords)
        : m_input(input), m_keywords(keywords)
    {
        readChar(); // Initialize the first character
    }

    void Lexer::readChar()
    {
        if (m_readPosition >= m_input.length())
        {
            m_char = 0; // EOF
        }
        else
        {
            m_char = m_input[m_readPosition];
        }
        m_position = m_readPosition;
        m_readPosition++;

        if (m_char == '\n')
        {
            m_line++;
            m_column = 0;
        }
        else
        {
            m_column++;
        }
    }

    char Lexer::peekChar() const
    {
        if (m_readPosition >= m_input.length())
        {
            return 0; // EOF
        }
        return m_input[m_readPosition];
    }

    void Lexer::skipWhitespace()
    {
        while (isspace(m_char))
        {
            readChar();
        }
    }

    void Lexer::skipSingleLineComment()
    {
        while (m_char != '\n' && m_char != 0)
        {
            readChar();
        }
    }

    void Lexer::skipMultiLineComment()
    {
        while (true)
        {
            if (m_char == '*' && peekChar() == '/')
            {
                readChar(); // consume '*'
                readChar(); // consume '/'
                break;
            }
            if (m_char == 0)
            {
                break;
            }
            readChar();
        }
    }

    bool Lexer::isLetter(char ch)
    {
        return isalpha(ch) || ch == '_';
    }

    std::string Lexer::readIdentifier()
    {
        size_t startPosition = m_position;
        while (isLetter(m_char) || isdigit(m_char) || m_char == '-')
        {
            readChar();
        }
        return m_input.substr(startPosition, m_position - startPosition);
    }

    // This function will need to be adapted for fully dynamic keywords.
    // For now, we'll keep it simple and assume the dynamic keywords map
    // will contain the full string like "[Template]".
    Token Lexer::readBlockKeyword()
    {
        int startLine = m_line;
        int startColumn = m_column;

        readChar(); // Consume '['
        std::string identifier = readIdentifier();
        if (m_char != ']')
        {
            return {TokenType::ILLEGAL, "[" + identifier, startLine, startColumn};
        }
        readChar(); // Consume ']'

        std::string full_block = "[" + identifier + "]";
        auto it = m_keywords.find(full_block);
        if (it != m_keywords.end())
        {
            Token tok = it->second;
            tok.line = startLine;
            tok.column = startColumn;
            return tok;
        }

        return {TokenType::ILLEGAL, full_block, startLine, startColumn};
    }

    std::string Lexer::readString(char quote_type)
    {
        size_t startPosition = m_position + 1;
        do
        {
            readChar();
        } while (m_char != quote_type && m_char != 0);

        std::string result = m_input.substr(startPosition, m_position - startPosition);
        readChar(); // consume closing quote
        return result;
    }

    std::string Lexer::readNumber()
    {
        size_t startPosition = m_position;
        while (isdigit(m_char) || m_char == '.')
        {
            readChar();
        }
        return m_input.substr(startPosition, m_position - startPosition);
    }

    std::string Lexer::readComment()
    {
        size_t startPosition = m_position;
        while (m_char != '\n' && m_char != 0)
        {
            readChar();
        }
        return m_input.substr(startPosition, m_position - startPosition);
    }

    Token Lexer::NextToken()
    {
        Token tok;
        skipWhitespace();

        tok.line = m_line;
        tok.column = m_column;

        switch (m_char)
        {
            case '=':
            case ':':
                tok = {TokenType::COLON, std::string(1, m_char), tok.line, tok.column};
                break;
            case ';':
                tok = {TokenType::SEMICOLON, std::string(1, m_char), tok.line, tok.column};
                break;
            case '+':
                tok = {TokenType::PLUS, std::string(1, m_char), tok.line, tok.column};
                break;
            case '-':
                tok = {TokenType::MINUS, std::string(1, m_char), tok.line, tok.column};
                break;
            case '*':
                if (peekChar() == '*') {
                    readChar(); // consume first '*'
                    tok = {TokenType::POWER, "**", tok.line, tok.column};
                } else {
                    tok = {TokenType::ASTERISK, std::string(1, m_char), tok.line, tok.column};
                }
                break;
            case '%':
                tok = {TokenType::MODULO, std::string(1, m_char), tok.line, tok.column};
                break;
            case '>':
                tok = {TokenType::GT, std::string(1, m_char), tok.line, tok.column};
                break;
            case '<':
                tok = {TokenType::LT, std::string(1, m_char), tok.line, tok.column};
                break;
            case '?':
                tok = {TokenType::QUESTION, std::string(1, m_char), tok.line, tok.column};
                break;
            case ',':
                tok = {TokenType::COMMA, std::string(1, m_char), tok.line, tok.column};
                break;
            case '/':
                if (peekChar() == '/') {
                    readChar(); readChar();
                    skipSingleLineComment();
                    return NextToken();
                } else if (peekChar() == '*') {
                    readChar(); readChar();
                    skipMultiLineComment();
                    return NextToken();
                } else {
                    tok = {TokenType::SLASH, std::string(1, m_char), tok.line, tok.column};
                }
                break;
            case '(': tok = {TokenType::LPAREN, "(", tok.line, tok.column}; break;
            case ')': tok = {TokenType::RPAREN, ")", tok.line, tok.column}; break;
            case '{': tok = {TokenType::LBRACE, "{", tok.line, tok.column}; break;
            case '}': tok = {TokenType::RBRACE, "}", tok.line, tok.column}; break;
            case '[':
                if (isLetter(peekChar())) {
                    return readBlockKeyword();
                } else {
                    tok = {TokenType::LBRACKET, "[", tok.line, tok.column};
                }
                break;
            case ']': tok = {TokenType::RBRACKET, "]", tok.line, tok.column}; break;
            case '@': tok = {TokenType::AT, "@", tok.line, tok.column}; break;
            case '.': tok = {TokenType::DOT, ".", tok.line, tok.column}; break;
            case '&': tok = {TokenType::AMPERSAND, "&", tok.line, tok.column}; break;
            case '"':
            case '\'':
                tok.type = TokenType::STRING;
                tok.literal = readString(m_char);
                return tok;
            case '#':
                if (peekChar() == ' ') {
                    readChar(); readChar();
                    tok.type = TokenType::COMMENT;
                    tok.literal = readComment();
                    return tok;
                } else {
                    tok = {TokenType::ILLEGAL, "#", tok.line, tok.column};
                }
                break;
            case 0:
                tok.type = TokenType::END_OF_FILE;
                tok.literal = "";
                return tok;
            default:
                if (isdigit(m_char)) {
                    tok.type = TokenType::NUMBER;
                    tok.literal = readNumber();
                    return tok;
                } else if (isLetter(m_char)) {
                    tok.literal = readIdentifier();
                    auto it = m_keywords.find(tok.literal);
                    if (it != m_keywords.end()) {
                        tok.type = it->second.type;
                    } else {
                        tok.type = TokenType::IDENT;
                    }
                    return tok;
                } else {
                    tok = {TokenType::ILLEGAL, std::string(1, m_char), tok.line, tok.column};
                }
                break;
        }

        readChar();
        return tok;
    }

    std::string Lexer::readRawBlockContent()
    {
        size_t start_pos = m_position;
        size_t scan_pos = m_position;
        int brace_level = 1;

        if (m_char == '{') brace_level++;
        if (m_char == '}') brace_level--;
        scan_pos = m_readPosition;

        while (brace_level > 0 && scan_pos < m_input.length())
        {
            char current_char = m_input[scan_pos];
            if (current_char == '{') brace_level++;
            else if (current_char == '}') brace_level--;
            scan_pos++;
        }

        size_t end_pos = (brace_level > 0) ? m_input.length() : scan_pos - 1;
        std::string content = (end_pos > start_pos) ? m_input.substr(start_pos, end_pos - start_pos) : "";

        while(m_position < end_pos) {
            readChar();
        }

        return content;
    }

} // namespace CHTL
